### PR TITLE
Feature: Moveable depots #6328

### PR DESCRIPTION
**Pull request by Eddi-z**
patch imported from flyspray
minor conflict resolved
savegame bump rebased ~~(TODO: put correct version info)~~ now resolved with new saveload enum
~~patch queue seems "unclean" (partial patch did not compile, whole queue does)~~ now resolved by reordering some patches

### CODE REVIEW

**Comment by andythenorth** [comment]
Gave this 5 minutes of testing, initial impressions are that all worked as expected.  Vehicle orders are preserved for up to 1 month, then invalidated if the depot isn't rebuilt by then.  The search radius for 'preserve depot' seems to be about 8 tiles or so.   In use, it's pretty neat.

FWIW, adf88 considered there was more to do on the original patch https://github.com/OpenTTD/OpenTTD/issues/6328#issuecomment-379454447 

However it might be useful as is.

**Comment by andythenorth** [comment]
From CI: 2019-01-12T22:39:23.2400344Z Running ai/regression/tst_regression... failed!

**Comment by glx22** [comment]
First regression error is at main.nut:1062, it's because the rail depot list is empty but it shouldn't as a rail depot as been built. 

Second regression error is at main.nut:1167, it's because the road depot list actually contains the water depot list.

File: src/script/api/script_depotlist.cpp [hunk 1]
```diff
@@ -40,9 +40,13 @@ ScriptDepotList::ScriptDepotList(ScriptTile::TransportType transport_type)
...
 
+	assert_compile(VEH_TRAIN == (int)ScriptTile::TRANSPORT_RAIL);
+	assert_compile(VEH_ROAD  == (int)ScriptTile::TRANSPORT_ROAD);
+	assert_compile(VEH_SHIP  == (int)ScriptTile::TRANSPORT_WATER);
+
 	/* Handle 'standard' depots. */
 	const Depot *depot;
 	FOR_ALL_DEPOTS(depot) {
-		if ((::GetTileOwner(depot->xy) == ScriptObject::GetCompany() || ScriptObject::GetCompany() == OWNER_DEITY) && ::IsTileType(depot->xy, tile_type)) this->AddItem(depot->xy);
+		if (depot->IsInUse() && (depot->owner == ScriptObject::GetCompany() || ScriptObject::GetCompany() == OWNER_DEITY) && depot->type == (int)tile_type) this->AddItem(depot->xy);
```
Comment (glx22) [inline src/script/api/script_depotlist.cpp]: Regression fails here

Comment (glx22) [reply src/script/api/script_depotlist.cpp]: Easy to fix, tile_type should be transport_type, and tile_type variable can be removed

File: src/order_cmd.cpp [hunk 2]
```diff
@@ -2092,7 +2078,7 @@ bool UpdateOrderDest(Vehicle *v, const Order *order, int conditional_depth, bool
 				v->IncrementRealOrderIndex();
 			} else {
 				if (v->type != VEH_AIRCRAFT) {
-					v->SetDestTile(Depot::Get(order->GetDestination())->xy);
+					v->SetDestTile(v->GetOrderDepotLocation(order->GetDestination()));
 				} else {
```
Comment (Eddi-z) [inline src/order_cmd.cpp:2048]: i'm not 100% sure i correctly resolved conflicts with 3e0e3cfac1 and 81330b8d6e here

**Comment by PeterN** [comment]
@Eddi-z will you update this?

**Comment by Eddi-z** [comment]
So, i resolved the saveload part, but the viewport part is way over my head with the kdtree changes.

**Comment by nielsmh** [comment]
I rebased this branch and updated it to use Kdtree here: https://github.com/nielsmh/OpenTTD/tree/movable-depots
Since I rebased I can't make a PR, and I don't want to push to Eddi's branch without an ok first.

Update: Got an ok on IRC.

File: src/depot.cpp [hunk 3]
```diff
@@ -31,21 +33,61 @@ Depot::~Depot()
...
+	OrderBackup::Reset(this->xy, false);
+
+	/* Mark that the depot is demolished and start the countdown. */
+	this->delete_ctr = 8;
+
+	/* Make sure no vehicle is going to the old depot. */
+	Vehicle *v;
+	FOR_ALL_VEHICLES(v) {
+		if (v->dest_tile == this->xy && v->First() == v && v->current_order.IsType(OT_GOTO_DEPOT)) {
+			v->dest_tile = v->GetOrderDepotLocation(this->index);
```
Comment (PeterN) [inline src/depot.cpp]: Use `v->SetDestTile()` so any caches are cleared.

File: src/depot_cmd.cpp [hunk 4]
```diff
@@ -78,8 +110,45 @@ CommandCost CmdRenameDepot(TileIndex tile, DoCommandFlag flags, uint32 p1, uint3
...
+}
+
+void OnTick_Depot()
+{
+	if (_game_mode == GM_EDITOR) return;
+
+	/* Clean up demolished depots. */
+	Depot *d;
+	FOR_ALL_DEPOTS(d) {
+		if ((_tick_counter + d->index) % DEPOT_REMOVAL_TICKS == 0) {
```
Comment (PeterN) [inline src/depot_cmd.cpp]: Duff this up to reduce indentation? `if (... != 0) continue;`

**Review by PeterN — CHANGES_REQUESTED** [review]

**Comment by PeterN** [comment]
Found a crash:
1. Build a depot
2. Build an engine
3. Sell the engine
4. Bulldoze the depot
5. Profit?

Something to do with Order backups...

**Comment by LordAro** [comment]
Hi, there's been no activity on this for over 4 months. If no more effort will be put into this PR, it will be closed in the next week or so

**Comment by andythenorth** [comment]
I think we can let this one slide into the darkness.  Moving depots is a nice feature if you use escape depots or similar, but it's not essential.

**Pull request closed without merging**